cpp/src/arrow/record_batch_test.cc

```diff
@@ -388,6 +389,34 @@ class TestRecordBatchReader : public ::testing::Test {
   std::shared_ptr<RecordBatchReader> reader_;
 };
 
+// A minimal class to test Status::Warn()
+class UncloseableReader : public RecordBatchReader {
+ public:
+  UncloseableReader(RecordBatchVector batches, std::shared_ptr<Schema> schema)
+      : schema_(std::move(schema)), it_(MakeVectorIterator(std::move(batches))) {}
+
+  ~UncloseableReader() { ARROW_WARN_NOT_OK(Close(), "Expect: uncloseable reader"); }
+
+  Status ReadNext(std::shared_ptr<RecordBatch>* batch) override {
+    return it_.Next().Value(batch);
+  }
+
+  std::shared_ptr<Schema> schema() const override { return schema_; }
+
+  Status Close() override { return Status::Invalid("uncloseable reader"); }
+
+ protected:
+  std::shared_ptr<Schema> schema_;
+  RecordBatchIterator it_;
+};
+
+TEST_F(TestRecordBatchReader, CloseAndWarn) {
+  auto uncloseable_reader_ =
+      std::make_shared<UncloseableReader>(batches_, reader_->schema());
+
+  ASSERT_EQ(Status::Invalid("uncloseable reader"), uncloseable_reader_->Close());
+}
```

Comment:
I'm not so sure this test is all that useful? If we just want to have coverage of Status::Warn, why not just call it directly from a test?